x/emissions/keeper/keeper.go

```diff
@@ -1139,44 +1161,205 @@ func (k *Keeper) SetDelegateStakeUponReputer(ctx context.Context, topicId TopicI
+		return err
+	}
+	byActorKey := collections.Join3(removalInfo.Reputer, removalInfo.TopicId, removalInfo.BlockRemovalCompleted)
+	return k.stakeRemovalsByActor.Set(ctx, byActorKey)
 }
 
 // remove a stake removal from the queue
-func (k *Keeper) DeleteStakeRemoval(ctx context.Context, topicId TopicId, address ActorId) error {
-	key := collections.Join(topicId, address)
-	return k.stakeRemoval.Remove(ctx, key)
+func (k *Keeper) DeleteStakeRemoval(
+	ctx context.Context,
+	blockHeight BlockHeight,
+	topicId TopicId,
+	address ActorId,
+) error {
+	byBlockKey := collections.Join3(blockHeight, topicId, address)
+	has, err := k.stakeRemovalsByBlock.Has(ctx, byBlockKey)
+	if err != nil {
+		return err
+	}
+	if !has {
+		return types.ErrStakeRemovalNotFound
+	}
+	err = k.stakeRemovalsByBlock.Remove(ctx, byBlockKey)
+	if err != nil {
+		return err
+	}
+	byActorKey := collections.Join3(address, topicId, blockHeight)
+	return k.stakeRemovalsByActor.Remove(ctx, byActorKey)
 }
 
-// For a given topic id and reputer address, get their stake removal information
-func (k *Keeper) GetDelegateStakeRemovalByTopicAndAddress(ctx context.Context, topicId TopicId, reputer ActorId, delegator ActorId) (types.DelegateStakePlacement, error) {
-	key := collections.Join3(topicId, reputer, delegator)
-	return k.delegateStakeRemoval.Get(ctx, key)
+// get info about a removal
+func (k Keeper) GetStakeRemoval(
+	ctx context.Context,
+	BlockHeight int64,
+	topicId TopicId,
+	reputer ActorId,
+) (types.StakeRemovalInfo, error) {
+	return k.stakeRemovalsByBlock.Get(ctx, collections.Join3(BlockHeight, topicId, reputer))
+}
+
+// get a list of stake removals for this block
+func (k *Keeper) GetStakeRemovalsForBlock(
+	ctx context.Context,
+	blockHeight BlockHeight,
+) ([]types.StakeRemovalInfo, error) {
+	ret := make([]types.StakeRemovalInfo, 0)
+	rng := collections.NewPrefixedTripleRange[BlockHeight, TopicId, ActorId](blockHeight)
+	iter, err := k.stakeRemovalsByBlock.Iterate(ctx, rng)
+	if err != nil {
+		return ret, err
+	}
+	for ; iter.Valid(); iter.Next() {
+		val, err := iter.Value()
+		if err != nil {
+			return ret, err
+		}
+		ret = append(ret, val)
+	}
+	return ret, nil
+}
+
+// get the first found stake removal for a reputer and topicId or err not found if not found
+func (k *Keeper) GetFirstStakeRemovalForReputerAndTopicId(
+	ctx sdk.Context,
+	reputer string,
+	topicId uint64,
+) (removal types.StakeRemovalInfo, found bool, err error) {
+	rng := collections.NewSuperPrefixedTripleRange[ActorId, TopicId, BlockHeight](reputer, topicId)
+	iter, err := k.stakeRemovalsByActor.Iterate(ctx, rng)
+	if err != nil {
+		return types.StakeRemovalInfo{}, false, err
+	}
+	keys, err := iter.Keys()
+	if err != nil {
+		return types.StakeRemovalInfo{}, false, err
+	}
+	keysLen := len(keys)
+	if keysLen == 0 {
+		return types.StakeRemovalInfo{}, false, nil
+	} else {
+		if keysLen > 1 {
+			ctx.Logger().Warn("Invariant failure! More than one stake removal found for reputer and topicId")
```

Comment:
ok cool

Reply:
if we do get into this state though, will this prevent this reputer from ever removing their stake or something similar?

Reply:
I believe based on the way the code is written, if the invariant is broken, this will allow someone to remove stake more quickly than the delay would have otherwise permitted.

The alternative would be to prevent someone from ever removing stake which seems maybe worse?

Reply:
as long as its not possible to lock them out of taking an action they should be able to take

Reply:
oh oops I ddin't realize that this doesn't return an error. Should it though, since it's an "invariant" ?

Reply:
actually we should be able to produce an error here you are right, because the CancelStake operation is exposed

so if the invariant is broken they should be able to cancel their stake until the invariant is no longer broken LOL

BUT BETTER TO NOT HAVE THE INVARIANT EVER BREAK

Reply:
I put up a PR that now makes the operation return an error

in the case of an error the cancel stake function now continues to let you cancel even if this specific error is observed